lib/cylc/xtriggers/xrandom.py

```diff
@@ -38,13 +38,47 @@ def xrandom(percent, secs=0, _=None, debug=False):
     The '_' argument is not used in the function code, but can be used to
     specialize the function signature to cycle point or task.
 
+    If the percent is zero, it returns that the trigger condition was
+    not satisfied, and an empty dictionary.
+    >>> xrandom(0, 0)
+    (False, {})
+
+    If the percent is not zero, but the random percent success is not met,
+    then it also returns that the trigger condition was not satisfied,
+    and an empty dictionary.
+    >>> import sys
+    >>> mocked_randint = lambda x, y: 10
+    >>> sys.modules[__name__].randint = mocked_randint
+    >>> xrandom(20, 0)
+    (False, {})
+
+    Finally, if the percent is not zero, and the random percent success is
+    met, then it returns that the trigger condition was satisfied, and a
+    dictionary containing random color and size as result.
+    >>> import sys
+    >>> mocked_randint = lambda x, y: 1
+    >>> sys.modules[__name__].randint = mocked_randint
+    >>> xrandom(1, 0)
+    (True, {'COLOR': 'orange', 'SIZE': 'small'})
+
+    Args:
+        percent (int): percent likelihood.
+        secs (int): seconds to sleep before starting the trigger.
+        _ (object): used to allow users to specialize the trigger
+                    with extra parameters.
+        debug (bool): flag to enable debug information.
+    Returns:
+        A tuple with the trigger flag (bool) which is set to True if the
+        trigger was evaluated successful, False otherwise, and a random
+        color and size (dict).
     """
     sleep(float(secs))
     results = {}
-    satisfied = (1 == randint(1, 100 / int(percent)))
+    satisfied = int(percent) != 0 and \
+        (1 == randint(1, 100 / int(percent)))  # nosec
```

Comment:
This breaks with `randint(1, 100/int(99))` because it  essentially you're giving it float `1< N < 1.1111...`..  it requires `a <= N <= b` with `a` and `b` both integers.

You could incorporate the percentage 0 by doing the following:
```
    satisfied =  randint(1, 100) <= int(percent)
```
(Or `<= int(percent) % 100`, to ensure value between 0-100)

Reply:
:open_mouth: ! Nice catch @dwsutherland ! I believe you found a new bug, and an interesting one!

And one that does not affect `7.8.x`, because in Python 2  `100/int(99)` is 1.

@dwsutherland , what if we simply change it to `(1 == randint(1, 100 // int(percent)))`? To maintain the old behaviour.

Reply:
```
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 100/int(99)
1
>>>
```

And

```
>>> from random import randint
>>> randint(1, 100/int(99))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/random.py", line 221, in randint
    return self.randrange(a, b+1)
  File "/usr/lib/python3.6/random.py", line 194, in randrange
    raise ValueError("non-integer stop for randrange()")
ValueError: non-integer stop for randrange()
```

Possibly would send another pull request, just to distinguish between this one that was backported to 7.8.x, and the different change for 8.0a1. How does that sound?

Reply:
Ouch!!! Good spotting.  In fact it breaks with most values of `percent`!  (My fault, in the original 😬 )

Reply:
Ahh.. I see you have to make the numerator a float in python2
```
>>> 100./int(99)
1.0101010101010102
```
The issue I have with;
`(1 == randint(1, 100 // int(percent)))`
is that you can't get a true success rate for 99 or even 90 .. because it gets truncated.

Reply:
Yes, my implementation is deficient, even in Python 2 - although it looks slightly less insane there at least.  It worked properly for my typical test cases, e.g. 10% ... but not in general!

Reply:
Then should we use 

`random.random() < int(percent)/100`

?

Fixes the `ZeroDivisionError` and the issue with `float` & `int`.